[BE][SparseAdam] cleaner way to verify no sparse params

janeyx99 opened the pull request:
## Deprecation
As of this PR, SparseAdam will become consistent with the rest of our optimizers in that it will only accept containers of Tensors/Parameters/param groups. Hitherto, the SparseAdam constructor had accidentally allow raw tensors as the params argument to the constructor. Now, if you write the following code, there will be a warning: "Passing in a raw Tensor as `params` to SparseAdam is deprecated. In the future, this will raise an error. Please wrap your Tensor in an iterable instead."

```
import torch
param = torch.rand(16, 32)
optimizer = torch.optim.SparseAdam(param)
```

Instead you should replace the last line with
```
optimizer = torch.optim.SparseAdam([param])
```
to avoid the warning.


## Context:

https://github.com/pytorch/pytorch/pull/47724 fixed the problem that SparseAdam could not handle generators by using the `list(...)` construct. However, this meant that SparseAdam deviated from other optimizers in that it could _accept_ a raw Tensors/Parameter vs requiring a container of them. This is not really a big deal.

So why this PR?

I do think this PR is cleaner. It uses the fact that the Optimizer parent class already containerizes parameters into parameter groups, so we could reuse that here by calling `super().__init__` first and then filter the param_groups after. This change would also make SparseAdam consistent with the rest of our optimizers in that only containerized params are accepted, which technically is BC breaking SO I've added a deprecation warning that we should remove in May 2024.

(But is it really BC breaking when we've said in the docs that params should be an iterable this whole time? Maybe this is just a bug fix....😛)

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #114178
* __->__ #114425

janeyx99 committed:
[BE][SparseAdam] cleaner way to verify no sparse params

[ghstack-poisoned]

drisspg commented on torch/optim/sparse_adam.py:
```diff
@@ -15,25 +15,21 @@ def __init__(self, params, lr=1e-3, betas=(0.9, 0.999), eps=1e-8, maximize: bool
-                for d_index, d_param in enumerate(param['params']):
-                    if d_param.is_sparse:
-                        sparse_params.append([index, d_index])
-            elif param.is_sparse:
-                sparse_params.append(index)
```
is this last branch handeld somewhere?

janeyx99 replied:
This shouldn't be, since the param groups we build have a certain structure. That's why there is the added assert in line 23.

drisspg reviewed (approved):
Sounds good, BC concerns on you lol

janeyx99 committed:
Update on "[BE][SparseAdam] cleaner way to verify no sparse params"

Context:

https://github.com/pytorch/pytorch/pull/47724 fixed the problem that SparseAdam could not handle generators by using the `list(...)` construct. However, this meant that SparseAdam deviated from other optimizers in that it could _accept_ a raw Tensors/Parameter vs requiring a container of them. This is not really a big deal.

So why this PR?

I do think this PR is cleaner. It uses the fact that the Optimizer parent class already containerizes parameters into parameter groups, so we could reuse that here by calling `super().__init__` first and then filter the param_groups after. This change would also make SparseAdam consistent with the rest of our optimizers in that only containerized params are accepted, which technically is BC breaking but would be minor, I believe.




[ghstack-poisoned]

janeyx99 committed:
Update on "[BE][SparseAdam] cleaner way to verify no sparse params"

Context:

https://github.com/pytorch/pytorch/pull/47724 fixed the problem that SparseAdam could not handle generators by using the `list(...)` construct. However, this meant that SparseAdam deviated from other optimizers in that it could _accept_ a raw Tensors/Parameter vs requiring a container of them. This is not really a big deal.

So why this PR?

I do think this PR is cleaner. It uses the fact that the Optimizer parent class already containerizes parameters into parameter groups, so we could reuse that here by calling `super().__init__` first and then filter the param_groups after. This change would also make SparseAdam consistent with the rest of our optimizers in that only containerized params are accepted, which technically is BC breaking but would be minor, I believe.




[ghstack-poisoned]

janeyx99 commented on torch/optim/optimizer.py:
```diff
@@ -255,9 +255,14 @@ def __init__(self, params: ParamsT, defaults: Dict[str, Any]) -> None:
-            raise TypeError("params argument given to the optimizer should be "
-                            "an iterable of Tensors or dicts, but got " +
-                            torch.typename(params))
+            if self.__class__.__name__ == 'SparseAdam':
```
happy to use a different method than looking into the dunders here, just didn't know of another way

janeyx99 commented:
Imported diff was all green, but due to the very very slight chance that someone may have used SparseAdam incorrectly between the last landed PR and now, I decided to be safe and add a deprecation warning.

mikaylagawarecki commented on torch/optim/optimizer.py:
```diff
@@ -255,9 +255,14 @@ def __init__(self, params: ParamsT, defaults: Dict[str, Any]) -> None:
+            if self.__class__.__name__ == 'SparseAdam':
+                warnings.warn(("Passing in a raw Tensor is deprecated. In the future, "
+                               "this will raise an error. Please wrap your Tensor in "
+                               "an iterable instead."), UserWarning)
```
nit
```suggestion
                               "an iterable instead."), FutureWarning)
```

janeyx99 replied:
What's the difference here?

mikaylagawarecki replied:
`FutureWarning` is a warning message that warns you about deprecated features that will be removed in a future version of a package.

I don't think we are that consistent about using `FutureWarning`, but I suppose it makes sense to do this in case users want to explicitly suppresss/allow certain warnings

mikaylagawarecki commented on torch/optim/optimizer.py:
```diff
@@ -255,9 +255,14 @@ def __init__(self, params: ParamsT, defaults: Dict[str, Any]) -> None:
-                            "an iterable of Tensors or dicts, but got " +
-                            torch.typename(params))
+            if self.__class__.__name__ == 'SparseAdam':
+                warnings.warn(("Passing in a raw Tensor is deprecated. In the future, "
```
nit: 
```suggestion
                warnings.warn(("Passing in a raw Tensor to SparseAdam is deprecated. In the future, "
```

janeyx99 committed:
Update on "[BE][SparseAdam] cleaner way to verify no sparse params"

Context:

https://github.com/pytorch/pytorch/pull/47724 fixed the problem that SparseAdam could not handle generators by using the `list(...)` construct. However, this meant that SparseAdam deviated from other optimizers in that it could _accept_ a raw Tensors/Parameter vs requiring a container of them. This is not really a big deal.

So why this PR?

I do think this PR is cleaner. It uses the fact that the Optimizer parent class already containerizes parameters into parameter groups, so we could reuse that here by calling `super().__init__` first and then filter the param_groups after. This change would also make SparseAdam consistent with the rest of our optimizers in that only containerized params are accepted, which technically is BC breaking SO I've added a deprecation warning that we should remove in May 2024.

(But is it really BC breaking when we've said in the docs that params should be an iterable this whole time? Maybe this is just a bug fix....😛)




[ghstack-poisoned]

janeyx99 commented:
@pytorchbot merge
